### PR TITLE
Beginning C++23: Chapter 17

### PR DESCRIPTION
**Pull request by Peter-VanWeert**

### CODE REVIEW

File: Exercises/Modules/Chapter 17/Soln17_01/Array.cppm [hunk 1]
```diff
@@ -1,45 +1,51 @@
...
-  ~Array();                                 // Destructor
-  Array(const Array& array);                // Copy constructor
-  Array& operator=(const Array& rhs);       // Copy assignment operator
-  void swap(Array& other) noexcept;         // Swap member function
-  T& operator[](size_t index);              // Subscript operator
-  const T& operator[](size_t index) const;  // Subscript operator-const arrays
-  void push_back(const T& element);         // Function to add new element to the end
-  size_t getSize() const { return m_size; } // Accessor for m_size
+  Array();                             // Default constructor
+  Array(std::size_t size);             // Constructor
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 17/Soln17_01/Array.cppm:11]: Was explicit before in the chapter.

File: Exercises/Modules/Chapter 17/Soln17_01/Soln17_01.cpp [hunk 2]
```diff
@@ -1,16 +1,16 @@
 // Adding a push_back() member function and a default constructor to the Array<> 
 // class template. Using copy-and-swap for a memory-safe push_back().
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 17/Soln17_01/Soln17_01.cpp:2]: exception-safe i.o. memory_safe?

File: Exercises/Modules/Chapter 17/Soln17_02/Soln17_02.cpp [hunk 3]
```diff
@@ -1,24 +1,23 @@
...
 
 int main()
 {
-  auto my_pair{ Pair<int, std::string>{122, "abc"} };
-  ++my_pair.first;
-  std::cout << "my_pair equals (" << my_pair.first 
-            << ", " << my_pair.second << ')' << std::endl;
-
-  auto pair1{ Pair<int, std::string>{0, "def"} };
+  auto pair = Pair<int, std::string>(122, "abc");
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 17/Soln17_02/Soln17_02.cpp:8]: uniform init?

File: Exercises/Modules/Chapter 17/Soln17_03/Soln17_03.cpp [hunk 4]
```diff
@@ -1,23 +1,55 @@
...
 
-  auto pair1{ Pair{  0, "def"s} };
-  auto pair2{ Pair{123, "abc"s} };
-  auto pair3{ Pair{123, "def"s} };
+    numbers[index] = generate_random_value();  // Store value at new index position
+  }
 
-  std::cout << (pair1 < pair2 && pair2 < pair3? "operator< seems to be working" : "oops") << std::endl;
-  std::cout << (pair1 == pair2? "oops" : "operator== works as well") << std::endl;
+  for (std::size_t i {}; i <= max_index; ++i)       // Create count entries in numbers array
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 17/Soln17_03/Soln17_03.cpp:50]: Wrong comment.

File: Exercises/Modules/Chapter 17/Soln17_04/LinkedList.cppm [hunk 5]
```diff
@@ -9,7 +9,7 @@ The latter has the implication that if you mis-use the container,
 your program will likely crash...
 */
 
-import <utility>;        // for std::swap()
+import std;
 
 export template<typename T>
 class LinkedList
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 17/Soln17_04/LinkedList.cppm:15]: the assignment operator is defined as:
`LinkedList& operator=(LinkedList list);`
which is not typical.
It usually is:
`LinkedList& operator=(const LinkedList& list);`
Also update its implementation to use the standard copy-and-swap idiom:
```
template<typename T>
LinkedList<T>& LinkedList<T>::operator=(const LinkedList& list)
{
	auto copy{ list };
	swap(copy);
	return *this;
}
```

File: Exercises/Modules/Chapter 17/Soln17_04/Soln17_04.cpp [hunk 6]
```diff
@@ -1,58 +1,52 @@
...
-// instead of a std::uniform_real_distribution to generate integers instead 
-// of float-point (or, "real") numbers.
-auto createUniformPseudoRandomNumberGenerator(int min, int max)
-{
-  std::random_device seeder;         // True random number generator to obtain a seed (slow)
-  std::default_random_engine generator{ seeder() };     // Efficient pseudo-random generator
-  std::uniform_int_distribution distribution{ min, max }; // Generate in [min, max) interval
-  return std::bind(distribution, generator);            //... and in the darkness bind them!
-}
+// Exercise 17-5 Exercising the LinkedList template class
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 17/Soln17_04/Soln17_04.cpp:1]: 17-4

File: Exercises/Modules/Chapter 17/Soln17_04/LinkedList.cppm (hunk 5 above)
Comment (MarcGregoire) [reply Exercises/Modules/Chapter 17/Soln17_04/LinkedList.cppm:15]: Same comment for LinkedList in Soln17_05

File: Examples/Modules/Chapter 17/Ex17_02/Ex17_02.cpp [hunk 7]
```diff
@@ -20,29 +17,28 @@ int main()
...
-        std::cout << (lines++ % 5 == 0 ? "\n" : "")
-                  << std::format("{:5g}", values[i] + values[i-1]);
+        if (lines++ % 5 == 0) std::println("");
+        std::print("{:5g}", values[i] + values[i-1]);
       }
     }
     catch (const std::out_of_range& ex)
     {
-      std::cerr << "\nout_of_range exception object caught! " << ex.what() << std::endl;
+      std::println("\nout_of_range exception object caught! {}", ex.what());
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 17/Ex17_02/Ex17_02.cpp:30]: I might have mentioned this before, but you can still use the error stream with std::print()/println().
For example:
`std::println(std::cerr, "\nout_of_range exception object caught! {}", ex.what());`

**Review by MarcGregoire — CHANGES_REQUESTED** [review]

File: Examples/Modules/Chapter 17/Ex17_02/Array.cppm [hunk 8]
```diff
@@ -1,37 +1,35 @@
...
-  explicit Array(size_t size);              // Constructor
+  explicit Array(std::size_t size);         // Constructor
   ~Array();                                 // Destructor
   Array(const Array& array);                // Copy constructor  
   Array& operator=(const Array& rhs);       // Assignment operator
   void swap(Array& other) noexcept;         // noexcept swap() function
   T& operator[](int index);                 // Subscript operator
   const T& operator[](int index) const;     // Subscript operator-const arrays
-  size_t getSize() const { return m_size; } // Accessor for size
+  std::size_t getSize() const noexcept { return m_size; } // Accessor for size
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 17/Ex17_02/Array.cppm:16]: There's too many of them. Nothing wrong with some variation.

File: Examples/Modules/Chapter 17/Ex17_04B/Stack.cppm [hunk 9]
```diff
@@ -35,13 +34,20 @@ private:
...
+// Conventional noexcept swap non-member function
+export template <typename T>
+void swap(Stack<T>& one, Stack<T>& other) noexcept
+{
+  one.swap(other);     // Forward to public member function
+}
+
 // Copy constructor
 template <typename T>
 Stack<T>::Stack(const Stack& stack)
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 17/Ex17_04B/Stack.cppm:53]: Interestingly, removing reset() even changes the behaviour for the better (if the destructor throws)...

File: Exercises/Modules/Chapter 17/Soln17_04/LinkedList.cppm (hunk 5 above)
Comment (Peter-VanWeert) [reply Exercises/Modules/Chapter 17/Soln17_04/LinkedList.cppm:15]: This variation is equally good. But I've added a comment acknowledging that. 

(As per https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Copy-and-swap it's even a better variation... Better for performance, better for combining move operator, ... Something to keep in mind for the next edition!)

File: Examples/Modules/Chapter 17/Ex17_02/Ex17_02.cpp (hunk 7 above)
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 17/Ex17_02/Ex17_02.cpp:30]: You have mentioned it several times before. And as I always: I know, but I don't cover `std::cerr` anymore in the book.

**Pull request merged**